File: app/javascript/channels/chatroom_channel.js
import consumer from "channels/consumer"

// const appChat = consumer.subscriptions.create("ChatroomChannel", {
const appChat = consumer.subscriptions.create({ channel: "ChatroomChannel", chatroom_id: 1 }, {
  connected() {
    console.log('connected');
  },

  disconnected() {
    console.log('disconnected');
  },

  received(data) {
    console.log(data);
    const messages = document.getElementById('messages');
    messages.insertAdjacentHTML('afterbegin', data['message']);
  },

  talk(data) {
    console.log(data);
    return this.perform('talk', { data: data });
  }
});

document.getElementById('content').addEventListener('keyup', function (e) {
  if (e.key === 'Enter') {
    const sender = document.getElementById('sender');
    appChat.talk({sender: sender.value, content: e.target.value});
    e.target.value = '';
    e.preventDefault();
  }
});

File: app/channels/chatroom_channel.rb
class ChatroomChannel < ApplicationCable::Channel
  def subscribed
    stream_from chatroom_name
  end

  def talk(data)
    message = Message.new
    message.published_at = Time.current
    message.sender = data.dig("data", "sender")
    message.content = data.dig("data", "content")
    message.save!

    # Messageモデルをレンダリングしてchatroom_channelへブロードキャストする
    ActionCable.server.broadcast chatroom_name, { message: render_message(message) }
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  private

  # Messageモデルをレンダリングするプライベートメソッド
  def render_message(message)
    ApplicationController.render(partial: "messages/message", locals: { message: message })
  end

  def chatroom_name
    "chatroom_#{params[:chatroom_id]}"
  end
end
